openc3/spec/models/widget_model_spec.rb

```diff
@@ -0,0 +1,182 @@
+      end
+
+      it "for the instance" do
+        model = WidgetModel.new(name: "DEFAULT", scope: 'scope')
+        expect(model).to be_a WidgetModel
+        model.handle_config(parser, 'DISABLE_ERB', ['aParm'])
+        expect{
+          model.handle_config(parser, 'NOT_A_KEYWORD', ['other', 'parms'])
+        }.to raise_error(OpenC3::ConfigParser::Error)
+      end
+    end
+
+    describe "self.names" do
+      it "returns all scope names" do
+        model = WidgetModel.new(name: "TEST", scope: 'DEFAULT')
+        model.create
+        model = WidgetModel.new(name: "OTHER", scope: 'DEFAULT')
+        model.create
+        names = WidgetModel.names(scope: 'DEFAULT')
+        # contain_exactly doesn't care about ordering and neither do we
+        expect(names).to contain_exactly("TEST", "OTHER")
+      end
+    end
+
+    describe "self.all" do
+      it "returns all the parsed scopes" do
+        model = WidgetModel.new(name: "DEFAULT", scope: 'scope')
+        model.create
+        model = WidgetModel.new(name: "OTHER", scope: 'scope')
+        model.create
+        all = WidgetModel.all(scope: 'scope')
+        expect(all.keys).to contain_exactly("DEFAULT", "OTHER")
+      end
+    end
+
+    describe "as_json" do
+      it "encodes all the input parameters" do
+        model = WidgetModel.new(name: "DEFAULT", scope: 'scope', updated_at: 12345)
+        json = model.as_json(:allow_nan => true)
+        expect(json['name']).to eql "DEFAULT"
+        expect(json['updated_at']).to eql 12345
+      end
+    end
+
+    describe "deploy" do
+      it "unless validate_only" do
+        dir = File.join(SPEC_DIR, "install")
+        model = WidgetModel.new(name: "DEFAULT", scope: 'scope', updated_at: 12346)
+        model.create
+        model.deploy(dir, {}, validate_only: true)
+        # not sure how to validate this
+      end
+
+      xit "the Widget's microservice" do
+        dir = File.join(SPEC_DIR, "install")
+        model = WidgetModel.new(name: "DEFAULT", scope: 'DEFAULT', updated_at: 12345)
+        model.create
+        model.deploy(dir, {})
+        # Ensure scope_model creates the UNKNOWN target and streams
+        target = TargetModel.get(name: "UNKNOWN", scope: "DEFAULT")
+        expect(target["name"]).to eql "UNKNOWN"
+        # and undeploy
+        model.undeploy
+        # nor how to validate undeployemnt
+      end
+    end
+
+    describe "destroy" do
+      it "destroys the scope and all microservices" do
+        s3 = instance_double("Aws::S3::Client")
+        allow(Aws::S3::Client).to receive(:new).and_return(s3)
+        options = OpenStruct.new
+        options.key = "blah"
+        objs = double("Object", :contents => [options], is_truncated: false)
+
+        scope = "TEST"
+        allow(s3).to receive(:put_bucket_policy)
+        allow(s3).to receive(:put_object)
+        allow(s3).to receive(:list_objects_v2).and_return(objs)
+        allow(s3).to receive(:delete_object).with(bucket: 'config', key: "blah")
```

Comment:
you can change allow to expect to ensure it is called

Reply:
Right, thanks.